comp/core/workloadmeta/collectors/internal/crio/crio.go

```diff
@@ -0,0 +1,299 @@
+	cpuQuota := float64(containerStatus.Resources.Linux.CpuQuota)
+	memLimitInBytes := uint64(containerStatus.Resources.Linux.MemoryLimitInBytes)
+
+	if cpuPeriod != 0 && cpuQuota != 0 {
+		limit := cpuQuota / cpuPeriod
+		cpuLimit = &limit
+	}
+	if memLimitInBytes != 0 {
+		memLimit = &memLimitInBytes
+	}
+	return cpuLimit, memLimit
+}
+
+// getContainerImage retrieves and converts a container image to workloadmeta format.
+func getContainerImage(ctx context.Context, client crio.ClientItf, imageSpec *v1.ImageSpec) workloadmeta.ContainerImage {
+	image, err := client.GetContainerImage(ctx, imageSpec)
+	if err != nil || image == nil {
+		log.Warnf("Failed to fetch image: %v", err)
+		return workloadmeta.ContainerImage{}
+	}
+
+	imgID := image.Id
+	imgName := ""
+	if len(image.RepoTags) > 0 {
+		imgName = image.RepoTags[0]
+	}
+	wmImg, err := workloadmeta.NewContainerImage(imgID, imgName)
+	if err != nil {
+		log.Warnf("Failed to create image: %v", err)
+		return workloadmeta.ContainerImage{}
+	}
+	if len(image.RepoDigests) > 0 {
+		wmImg.RepoDigest = image.RepoDigests[0]
+	}
+	return wmImg
+}
+
+// getContainerState returns the workloadmeta.ContainerState based on container status.
+func getContainerState(containerStatus *v1.ContainerStatus) workloadmeta.ContainerState {
+	if containerStatus == nil {
+		return workloadmeta.ContainerState{Status: workloadmeta.ContainerStatusUnknown}
+	}
+	exitCode := int64(containerStatus.ExitCode)
+	return workloadmeta.ContainerState{
+		Running:    containerStatus.State == v1.ContainerState_CONTAINER_RUNNING,
+		Status:     mapContainerStatus(containerStatus.State),
+		CreatedAt:  time.Unix(0, containerStatus.CreatedAt),
+		StartedAt:  time.Unix(0, containerStatus.StartedAt),
+		FinishedAt: time.Unix(0, containerStatus.FinishedAt),
+		ExitCode:   &exitCode,
+	}
+}
+
+// mapContainerStatus maps CRI-O container state to workloadmeta.ContainerStatus.
+func mapContainerStatus(state v1.ContainerState) workloadmeta.ContainerStatus {
+	switch state {
+	case v1.ContainerState_CONTAINER_CREATED:
+		return workloadmeta.ContainerStatusCreated
+	case v1.ContainerState_CONTAINER_RUNNING:
+		return workloadmeta.ContainerStatusRunning
+	case v1.ContainerState_CONTAINER_EXITED:
+		return workloadmeta.ContainerStatusStopped
+	case v1.ContainerState_CONTAINER_UNKNOWN:
+		return workloadmeta.ContainerStatusUnknown
+	}
+	return workloadmeta.ContainerStatusUnknown
+}
+
+// generateUnsetEvent creates an unset event for a given container ID.
+func generateUnsetEvent(seenID workloadmeta.EntityID) workloadmeta.CollectorEvent {
+	return workloadmeta.CollectorEvent{
+		Type:   workloadmeta.EventTypeUnset,
+		Source: workloadmeta.SourceRuntime,
+		Entity: &workloadmeta.Container{
+			EntityID: seenID,
+		},
+	}
+}
+
+// extractPortsFromAnnotations parses container ports from annotations.
+func extractPortsFromAnnotations(annotations map[string]string) []workloadmeta.ContainerPort {
```

Comment:
This way of getting container ports feels pretty hacky. The CRI-API does not expose container ports, but Kubernetes adds ports into the annotations for some of its pods like `coredns` and `kube-apiserver`. I added this in, but maybe it should just be removed.